### ISSUE TITLE
Scylla tries to delete missing sstable

### ISSUE THREAD

**Issue by duarten** [scylladb/scylla#1095]
## Installation details
Scylla version (or git commit hash): 0.19-20160308.4e3b98f-ub
Cluster size: 1
OS (RHEL/CentOS/Ubuntu/AWS AMI): Ubuntu 14.04, with XFS.

I got the following error by trying to run `nodetool compact` after running cassandra-test on a system that was forcefully rebooted:

```
scylla[1617]:  [shard 0] sstable - Exception when deleting sstable file: boost::filesystem::filesystem_error (boost::filesystem::canonical: No such file or directory: "/mnt/data/scylla/data/keyspace1/standard1-4df11f40f1c411e5b62a000000000000/keyspace1-standard1-ka-297-TOC.txt")
```
I have keyspace1-standard1-ka-117-TOC.txt on disk, then following that I have a bunch of keyspace1-standard1-ka-xxx-TOC.txt.tmp, including generation 297.

Issue might be related to marking non-sealed sstables for deletion.

**Comment by slivne** [comment]
@duarten the version is 0.19 it is not the rc1 version I sent in the email - please try and reproduce on the correct version.

**Comment by duarten** [comment]
Oh, sorry about that. Will try to repro.

**Comment by duarten** [comment]
Reproduced:

```
scylla[11308]:  [shard 0] sstable - Exception when deleting sstable file: boost::filesystem::filesystem_error (boost::filesystem::canonical: No such file or directory: "/mnt/data/scylla/data/keyspace1/standard1-4df11f40f1c411e5b62a000000000000/keyspace1-standard1-ka-121-TOC.txt")
scylla[11308]:  [shard 0] sstable - Exception when deleting sstable file: boost::filesystem::filesystem_error (boost::filesystem::canonical: No such file or directory: "/mnt/data/scylla/data/keyspace1/standard1-4df11f40f1c411e5b62a000000000000/keyspace1-standard1-ka-123-TOC.txt")
scylla[11308]:  [shard 0] sstable - Exception when deleting sstable file: boost::filesystem::filesystem_error (boost::filesystem::canonical: No such file or directory: "/mnt/data/scylla/data/keyspace1/standard1-4df11f40f1c411e5b62a000000000000/keyspace1-standard1-ka-126-TOC.txt")
scylla[11308]:  [shard 0] sstable - Exception when deleting sstable file: boost::filesystem::filesystem_error (boost::filesystem::canonical: No such file or directory: "/mnt/data/scylla/data/keyspace1/standard1-4df11f40f1c411e5b62a000000000000/keyspace1-standard1-ka-131-TOC.txt")
```

**Comment by raphaelsc** [comment]
@duarten: I got a slightly different problem, but I am confident that the problem is the same: WARN  2016-03-24 16:51:17,494 [shard 0] sstable - Exception when deleting sstable file: std::system_error (error system:2, No such file or directory)

I reproduced it by eating all disk  before running 'nodetool compact keyspace1', so compaction will fail when writing a new sstable, and this partial sstable will be marked for deletion. We probably shouldn't mark it for deletion because that's intended for sealed sstables. I think we need to implement a function that will delete partial sstables for a compaction that failed. @glommer?

**Comment by duarten** [comment]
(Version is 666.development-20160324.0afd1c6-ubuntu1)

**Comment by glommer** [comment]
@raphaelsc I think we should keep deleting compaction files. Compactions can increase the disk utilization twofold almost, so we can't leave failed compactions just lying around.

The message @duarten posted indicates that we're trying to delete the TOC. If the SSTable write fails, it's expected that the TOC is non-existent. Why aren't we more robust about it? I thought we could delete unfinished SSTables just fine.

**Comment by raphaelsc** [comment]
@glommer : I think that this commit should handle it: 4e52b41a46ca68daa3312a08db424ce0d2ca24c4

I see that it was applied just few hours ago.

**Comment by glommer** [comment]
Although that commit was done for another bug with different symptoms, it does look like it would handle this case.

**Comment by slivne** [comment]
Fix is merged into 1.0 - please retest with https://s3.amazonaws.com/downloads.scylladb.com/deb/unstable/ubuntu/branch-1.0/8/scylla.list

**Comment by duarten** [comment]
Tested and got this:

```
Mar 27 11:23:27 vagrant-ubuntu-trusty-64 scylla[1759]:  [shard 0] compaction_manager - compaction failed: read exception: seastar::broken_pipe_exception (Broken pipe), write_exception: std::system_error (error system:28, No space left on device)
Mar 27 11:23:27 vagrant-ubuntu-trusty-64 scylla[1759]:  [shard 0] compaction_manager - compaction task handler sleeping for 80 seconds
Mar 27 11:23:27 vagrant-ubuntu-trusty-64 scylla[1759]:  [shard 0] sstable - Exception when deleting sstable file: std::system_error (error system:2, No such file or directory)
```

There's still a ENOENT, but with less information.

**Comment by nyh** [comment]
On Sun, Mar 27, 2016 at 2:27 PM, Duarte Nunes <notifications@github.com>
wrote:

> Tested and got this:
>
> Mar 27 11:23:27 vagrant-ubuntu-trusty-64 scylla[1759]:  [shard 0] compaction_manager - compaction failed: read exception: seastar::broken_pipe_exception (Broken pipe), write_exception: std::system_error (error system:28, No space left on device)
> Mar 27 11:23:27 vagrant-ubuntu-trusty-64 scylla[1759]:  [shard 0] compaction_manager - compaction task handler sleeping for 80 seconds
> Mar 27 11:23:27 vagrant-ubuntu-trusty-64 scylla[1759]:  [shard 0] sstable - Exception when deleting sstable file: std::system_error (error system:2, No such file or directory)
>
> There's still a ENOENT, but with less information.
>

Looking at the the remove_by_toc_name() code it *almost* does the right
thing: deletes all the components listed in the toc, in parallel. If any of
these components fail to be deleted, one or more exceptions can be found by
parallel_for_each but only one will be reported (I'm tired of explaining
how wrong that is...).

So far *almost* so good - the code deleted all the existing files and warns
about those which don't. The problem is that the last two lines of
remove_by_toc_name() do not run, and the toc file itself got not get
deleted. This is probably wrong.

**Comment by avikivity** [comment]
Flow should be:
  - remove all components in parallel
    - forgive ENOENT, since the compoent may not have been written; otherwise deletion error should be raised
  - fsync the directory
  - delete the temporary TOC

**Comment by raphaelsc** [comment]
On Sun, Mar 27, 2016 at 9:17 AM, Avi Kivity <notifications@github.com>
wrote:

> Flow should be:
>
>    - remove all components in parallel
>       - forgive ENOENT, since the compoent may not have been written;
>       otherwise deletion error should be raised
>
> I sent a patch which consisted of skipping deletion if file doesn't exist,
but I think this approach is better. I will send a patch tomorrow.


>
>    -
>       - fsync the directory
>    - delete the temporary TOC
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/scylladb/scylla/issues/1095#issuecomment-202053645>
>



-- 
Raphael S. Carvalho
